### ISSUE TITLE
Add dropout

### ISSUE THREAD

**Issue by hughperkins** [hughperkins/DeepCL#4]
- dropout is a vital function, not currently implemented.  pretty vital for any self-respecting convolutional neural network really ;-)
- it doesnt need to be added to all propagate and backprop implementations, but probably at least to:
  - Propagate1: generic gpu-based forward propagation layer
  - BackpropErrorsv2Naive: generic gpu-based backprop of error gradient to previous layer
  - BackpropWeights2Naive: generic gpu-based backprop of error gradient onto weights of same layer
- you'd need to also add the Dropout option to ConvolutionalMaker class
- and you'd need to make all other implementations (Propagate2 etc...) throw a runtime_error, in their constructor, if dropout is requested in the passed-in maker object

**Comment by hughperkins** [comment]
Following discussion with Tambet, implementing this as a separate layer.  See https://github.com/hughperkins/DeepCL/tree/dropout

**Comment by hughperkins** [comment]
Done :-)
